core/src/main/java/org/apache/calcite/avatica/util/SecurityUtils.java

```diff
@@ -0,0 +1,266 @@
+      return lookup.findStatic(controllerKlass, "getContext",
+        MethodType.methodType(contextklass));
+    } catch (Throwable x) {
+      return null;
+    }
+  }
+
+  /**
+   * Get the current security manager, if available.
+   * @return the current security manager, if available, null otherwise
+   */
+  public static Object getSecurityManager() {
+    try {
+      // Use reflection to work with Java versions that have and don't have SecurityManager.
+      return System.class.getMethod("getSecurityManager").invoke(null);
+    } catch (Throwable ignored) {
+      return null;
+    }
+  }
+
+  /**
+   * <p>
+   * Checks the given permission, if the {@link #getSecurityManager() security manager} is set.
+   * </p>
+   * @param permission the permission to check
+   * @throws SecurityException if the permission check fails
+   */
+  public static void checkPermission(Permission permission) throws SecurityException {
+    Object securityManager = SecurityUtils.getSecurityManager();
+    if (securityManager == null) {
+      return;
+    }
+    try {
+      securityManager.getClass().getMethod("checkPermission").invoke(securityManager,
+          permission);
+    } catch (SecurityException x) {
+      throw x;
+    } catch (Throwable ignored) {
+    }
+  }
+
+  /**
+   * <p>
+   * Runs the given action with the calling context restricted to just the calling frame, not all
+   * the frames in the stack.
+   * </p>
+   * @param action the action to run
+   * @return the result of running the action
+   * @param <T> the type of the result
+   */
+  public static <T> T doPrivileged(PrivilegedAction<T> action) {
+    // Keep this method short and inlineable.
+    if (DO_PRIVILEGED == null) {
+      return action.run();
+    }
+    return doPrivileged(DO_PRIVILEGED, action);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <T> T doPrivileged(MethodHandle doPrivileged, PrivilegedAction<T> action) {
+    try {
+      return (T) doPrivileged.invoke(action);
+    } catch (RuntimeException | Error x) {
+      throw x;
+    } catch (Throwable x) {
+      // Unlikely to happen, as we catch Errors and RuntimeExceptions above, and doPrivileged
+      // does not throw checked exceptions
+      throw new RuntimeException(x);
+    }
+  }
+
+  /**
+   * <p>
+   * Runs the given action as the given subject.
+   * </p>
+   * @param subject the subject this action runs as
+   * @param action the action to run
+   * @return the result of the action
+   * @param <T> the type of the result
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T callAs(Subject subject, Callable<T> action) {
+    try {
+      if (CALL_AS == null) {
+        throw new RuntimeException(
+            "Was unable to run either of Subject.callAs() or Subject.doAs()");
+      }
+      return (T) CALL_AS.invoke(subject, action);
+    } catch (RuntimeException | Error x) {
+      throw x;
+    } catch (Throwable x) {
```

Comment:
The `CompletionException` is thrown directly... why do we rewrap it?

Reply:
Should be fixed by unwrapInvocationTargetException.

Reply:
unwrapInvocationTargetException is not needed, I will remove it.